File: Arrays/08_MoveZeroToEnd/Answer.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        int j = -1;

        // Find first zero
        for(int i = 0; i < nums.size(); i++)
        {
            if(nums[i] == 0)   // ✅ fixed (was nums[j])
            {
                j = i;
                break;
            }
        }

        if(j == -1)
        {
            return;
        }

        // Swap non-zero elements
        for(int i = j + 1; i < nums.size(); i++)
        {
            if(nums[i] != 0)
            {
                swap(nums[i], nums[j]);
                j++;
            }
        }
    }
};

// Time Complexity: O(n)
// Space Complexity: O(1)

int main() {
    int n;

    cout << "Enter number of elements: ";
    cin >> n;

    vector<int> nums(n);

    cout << "Enter elements: ";
    for(int i = 0; i < n; i++) {
        cin >> nums[i];
    }

    Solution obj;
    obj.moveZeroes(nums);

    cout << "Array after moving zeroes: ";
    for(int i = 0; i < n; i++) {
        cout << nums[i] << " ";
    }

    cout << endl;

    return 0;
}

File: Arrays/08_MoveZeroToEnd/Answer.java
import java.util.*;

// class Solution {
//     public void moveZeroes(int[] nums) {
        
        // BruteForce

        // Time Complexity: O(n)
        // Space Complexity: O(n) (extra array used)
        
//         ArrayList<Integer> temp = new ArrayList<>();

//         // Store non-zero elements
//         for(int i = 0; i < nums.length; i++)
//         {
//             if(nums[i] != 0)
//             {
//                 temp.add(nums[i]);
//             }
//         }

//         int nz = temp.size();

//         // Copy non-zero elements
//         for(int i = 0; i < nz; i++)
//         {
//             nums[i] = temp.get(i);
//         }

//         // Fill remaining with zeroes
//         for(int i = nz; i < nums.length; i++)
//         {
//             nums[i] = 0;
//         }
//     }
// }

class Solution {
    public void moveZeroes(int[] nums) {
        
        // OPTIMAL SOLUTION 

        // Time Complexity: O(n)
        // Space Complexity: O(1)

       int j = 0;

       for(int i = 0; i < nums.length; i++)
       {
            if(nums[i] != 0)
            {
                int temp = nums[i];
                nums[i] = nums[j];
                nums[j] = temp;
                j++;
            }
       }
    }
}

public class Answer {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number of elements: ");
        int n = sc.nextInt();

        int[] nums = new int[n];

        System.out.print("Enter elements: ");
        for(int i = 0; i < n; i++) {
            nums[i] = sc.nextInt();
        }

        Solution obj = new Solution();
        obj.moveZeroes(nums);

        System.out.print("Array after moving zeroes: ");
        for(int i = 0; i < n; i++) {
            System.out.print(nums[i] + " ");
        }

        sc.close();
    }
}